### PR TITLE
Migrate kustomize.go to Kustomize3

### PR DESCRIPTION
**Pull request by jbrette**
This PR is migrating the kustomize.go from kustomize 2.x to kustomize 3.x

- It is not really leveraging the new kustomize features yet.
- It is compiling with the new version of the kustomize v3 pkg.
- bases: field is deprecated in kustomize 3.x.  Still need to complete the merge the handling of the bases: and resources: fields in the kustomize.go

Corresponding manifest: [PR](https://github.com/kubeflow/manifests/pull/309). 

This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/4055)

### CODE REVIEW

**Comment by jbrette** [comment]
@kkasravi Please have a look at this [PR](https://github.com/kubernetes-sigs/kustomize/pull/1451). I let's the user specify a sorting order before running kubectl apply.  

With the current 3.1.0 kustomize, you can actually run `kustomize build . --reorder=none` in which case the kustomize.go in kubeflow repo may still be usefull.

Once the PR is merged, you will be able to add a simple kubectlapplyordertransformer.yaml in the kustomize folder...and the resources will be ordered accordingly.

Have a look at that [issue](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_0821) for an example.

The path of least resistance for you to try it is most likely to:
1, git clone -b allinone https://github.com/keleustes/kustomize.git
2. make install
3. make build-plugins

Note that the allinone branch is up to date with master but contains all the additional improvment PR our group proposed (diamond import, auto variable, inlining, ordering, fieldspec skipping).

/cc @ianhowell

**Comment by gabrielwen** [comment]
thanks for the PR!  overall LGTM, but would like to have second pair of eyes to make sure this is air tight :)

**Comment by kkasravi** [comment]
@gabrielwen @jbrette this was moved to the new repo kubeflow/kfctl#31
cc @jlewi 

Jerome - I would like to thank you again for providing this PR.
Your PR is an identical copy of kubeflow/kfctl#31 with 2 exceptions:
- The yaml changes under config aren't needed in the kfctl repo since these files are going under manifest
- A one line change in kustomize.go to keep Bases:
```
			parent.Bases = append(parent.Bases, basePath)
```
Until we have a solution from kustomize on multiple overlays that aren't required to differ by namespace, nameprefix or namesuffix per overlay the match PR that @jbrette opened which changes bases: -> resources: in manifests will break on configMapGenerator, patchesStrategicMerge and patchesJson6902. The resolution - which works fine in kustomize-v3.0.1 - is to move these definitions up into a common kustomization file. This is what kfctl has done all along. 


I've done some fairly extensive testing with kfctl using kustomize-v3.1.0

**Comment by jbrette** [comment]
@kkasravi Do you need me to abandon this PR since kfctl has been moved to its own repo. I have no problem to do so if this what is good for the project.

**Comment by kkasravi** [comment]
> @kkasravi Do you need me to abandon this PR since kfctl has been moved to its own repo. I have no problem to do so if this what is good for the project.

yes - please do. the kfctl repo has a script which syncs with kubeflow/bootstrap.
I picked up your changes when I was doing e2e / ci to see what obstacles kustomize-v3.1.0 presented. If you would like to open a new PR in kfctl - please do and we can merge that. 
Alternatively, I know how to change commit owners on a branch, so i could chown to you if 
that would be your preference. Or feel free to commit on the PR itself.

**Comment by kkasravi** [comment]
@jbrette btw - any update on kustomize accepting some of your work. 

BTW - https://github.com/kubernetes-sigs/kustomize/pull/1217 is genius.

**Comment by jbrette** [comment]
As advised, closed this PR since the content has been transferred to the kfctl repo.

**Comment by jbrette** [comment]
@kkasravi As for the kustomize work, the group already merged quite a lot of PR but mainly bug fixes one. It takes more time to get new "features" related PR merged.  But if the community continues to show support, then ultimetaly.

Meanwhile our project is using all those PR. We merged them into the [allinone branch](https://github.com/keleustes/kustomize)
 
Our kustomize usage is quite extensiv: [check here](https://github.com/keleustes/treasuremap).

**Comment by kkasravi** [comment]
thanks - lots of PatchesStrategicMerge - looks like an interesting project.

**Comment by kkasravi** [comment]
@jbrette 
waiting on https://github.com/kubeflow/manifests/pull/355 which in turn is waiting on https://github.com/kubeflow/manifests/pull/323 to be reverted

File: bootstrap/config/overlays/application/kustomization.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
-bases:
+resources:
```
[the remark this answers is not in the document]
Comment (jbrette) [reply bootstrap/config/overlays/application/kustomization.yaml]: done

File: bootstrap/config/overlays/basic_auth/kustomization.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
-bases:
+resources:
```
[the remark this answers is not in the document]
Comment (jbrette) [reply bootstrap/config/overlays/basic_auth/kustomization.yaml]: done

File: bootstrap/config/overlays/gcp/kustomization.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
-bases:
+resources:
```
[the remark this answers is not in the document]
Comment (jbrette) [reply bootstrap/config/overlays/gcp/kustomization.yaml]: done

File: bootstrap/config/overlays/aws/kustomization.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
-bases:
+resources:
```
[the remark this answers is not in the document]
Comment (jbrette) [reply bootstrap/config/overlays/aws/kustomization.yaml]: done

File: bootstrap/config/overlays/istio/kustomization.yaml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
-bases:
+resources:
```
[the remark this answers is not in the document]
Comment (jbrette) [reply bootstrap/config/overlays/istio/kustomization.yaml]: done

File: bootstrap/pkg/kfapp/kustomize/kustomize.go [hunk 6]
```diff
@@ -743,7 +746,7 @@ func MergeKustomization(compDir string, targetDir string, kfDef *kfdefsv3.KfDef,
 	if child.Bases == nil {
 		basePath := extractSuffix(compDir, targetDir)
 		if _, ok := kustomizationMaps[basesMap][basePath]; !ok {
-			parent.Bases = append(parent.Bases, basePath)
+			parent.Resources = append(parent.Resources, basePath)
```
[the remark this answers is not in the document]
Comment (jbrette) [reply bootstrap/pkg/kfapp/kustomize/kustomize.go]: done

File: bootstrap/pkg/kfapp/kustomize/kustomize.go [hunk 7]
```diff
@@ -752,7 +755,7 @@ func MergeKustomization(compDir string, targetDir string, kfDef *kfdefsv3.KfDef,
 		baseAbsolutePath := path.Join(targetDir, value)
 		basePath := extractSuffix(compDir, baseAbsolutePath)
 		if _, ok := kustomizationMaps[basesMap][basePath]; !ok {
-			parent.Bases = append(parent.Bases, basePath)
+			parent.Resources = append(parent.Resources, basePath)
```
[the remark this answers is not in the document]
Comment (jbrette) [reply bootstrap/pkg/kfapp/kustomize/kustomize.go]: done

File: components/profile-controller/config/default/kustomization.yaml [hunk 8]
```diff
@@ -17,15 +20,15 @@ commonLabels:
 # files that kustomize reads, modifies and emits as a
 # YAML string, with resources separated by document
 # markers ("---").
-bases:
+resources:
```
[the remark this answers is not in the document]
Comment (jbrette) [reply components/profile-controller/config/default/kustomization.yaml]: done

File: releasing/auto-update/overlays/cron/kustomization.yaml [hunk 9]
```diff
@@ -1,7 +1,8 @@
 namespace: kf-releasing
-bases:
-  - ../../base
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
 resources:
+  - ../../base
```
[the remark this answers is not in the document]
Comment (jbrette) [reply releasing/auto-update/overlays/cron/kustomization.yaml]: done

File: releasing/auto-update/overlays/job/kustomization.yaml [hunk 10]
```diff
@@ -1,9 +1,10 @@
 namespace: kf-releasing
-bases:
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
   - ../../base
-resources:
```
[the remark this answers is not in the document]
Comment (jbrette) [reply releasing/auto-update/overlays/job/kustomization.yaml]: done

**Comment by kkasravi** [comment]
perfect, thanks @jbrette 

/lgtm

**Comment by jbrette** [comment]
> "sigs.k8s.io/kustomize/v3/plugin/builtin"

done. I've got questions:
1. Where in the logs did find the origin of the issue.
2. Have a look at that code: https://github.com/keleustes/kustomize/tree/order/plugin/kustomize.config.k8s.io/v1beta1/kubectlapplyordertransformer and that [PR](https://github.com/kubernetes-sigs/kustomize/pull/1451).
It let's you control the order of the resources (in particular the CRD) during kubectl apply and kubectl delete.

**Comment by kkasravi** [comment]
thanks @jbrette 

I found the order issue when I had merged kubeflow/manifests/application/application-crds and kubeflow/manifests/application/application. Under this scenario the k8s server complained that the crd wasn't known (applications.app.k8s.io). I kept the two as is (application-crds, application) and used the legacy transformer. This seemed to fix the problem - although I can't understand why I would get this error if the application crd is first and the application is later in the same yaml

**Comment by kkasravi** [comment]
this is still waiting on kubeflow/manifests#355

**Comment by kkasravi** [comment]
/assign @jlewi 

update kfctl to kustomize-v3.0.1

**Comment by jlewi** [comment]
@kkasravi @jbrette What is the backwards compatibility story?In particular what is the impact on our ability to upgrade Kubeflow deployments from v0.6 to v0.7.

It looks like the E2E tests in this PR are passing; so does this mean the updated version of kctl can still apply the old style kustomize manifests?

@kkasravi was there something in particular you wanted me to look at it? If not you can just go ahead and approve it

**Comment by kkasravi** [comment]
> It looks like the E2E tests in this PR are passing; so does this mean the updated version of kctl can still apply the old style kustomize manifests?
yes


> @kkasravi @jbrette What is the backwards compatibility story?In particular what is the impact on our ability to upgrade Kubeflow deployments from v0.6 to v0.7.

one change was required in manifests kubeflow/manifests#355 to upgrade.
Current upgrade functionality that @richardsliu is authoring should work - since he ran it with me recently. 



> @kkasravi was there something in particular you wanted me to look at it? If not you can just go ahead and approve it

I tried approving - the label didn't take - not sure why. I'll try again
/approve

**Comment by kkasravi** [comment]
oh - I'm not listed in the OWNERS file - https://github.com/kubeflow/kubeflow/blob/master/OWNERS

**Comment by jbrette** [comment]
@kkasravi @jlewi Do I need to do something. Do you need help for something before being able to merge ?

**Comment by kkasravi** [comment]
I think @jlewi needs to approve

**Comment by jbrette** [comment]
@jlewi @kkasravi Was forced to rebase to pick today's merge related to jsonnet.py. You will need to reapply the /lgtm tag again.

**Pull request merged**